### PR TITLE
Two dimensional turbulence shallow water example

### PR DESCRIPTION
**Pull request by francispoulin**
Inspired by @Lichriszz and their problem in [#1362](https://github.com/CliMA/Oceananigans.jl/discussions/1362#discussioncomment-374286) I started to make an example of two dimensional turbulence using the shallow water model.  At the moment the results are not necessarily worth looking at but I wonder if people could look at the code [here](https://github.com/CliMA/Oceananigans.jl/blob/fjp/two-dimensional-turbulence-shallow-water-example/examples/two_dimensional_turbulence_shallow_water.jl) and say whether the set up is what we want?

A few issues that I have come cross are the following:

- [x] Use the time stepping wizard for numerical stability
- [x] Reduce the number of import statements
- [ ] Plot the vorticity and the divergence fields
- [x] Pick the parameters to get nice results.  What is a good a reference for a planar geometry?
- [ ] Should test on GPU
- [ ] Should test with higher resolution.

One possible paper is [Polvani et al (1994)](https://aip.scitation.org/doi/pdf/10.1063/1.166002)

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
I think the setup looks great! Hopefully we can reduce the number of import statements (especially after #1132).

The 2D turbulence example always looks great but it might be nice to do something different just to have more variety in the examples. Is there a neat example that requires a shallow water model?

Alternatively we could convert the 2D turbulence example to use `ShallowWaterModel` if it doesn't change the output much. This would increase variety among the existing examples.

Having an example with `ShallowWaterModel` would suggest it's ready for use though, so it should then be exported.

**Comment by francispoulin** [comment]
> I think the setup looks great! Hopefully we can reduce the number of import statements (especially after #1132).
> 
> The 2D turbulence example always looks great but it might be nice to do something different just to have more variety in the examples. Is there a neat example that requires a shallow water model?
> 
> Alternatively we could convert the 2D turbulence example to use `ShallowWaterModel` if it doesn't change the output much. This would increase variety among the existing examples.
> 
> Having an example with `ShallowWaterModel` would suggest it's ready for use though, so it should then be exported.

Thanks @ali, for the quick and thoughtful reply.  Some excellent points.

Inspired by your comments, I added two bullets above.

First, reduce the number of import statements.  This because I did not import as much elsewhere, and I agree this should be fixed to make ti easier to use.

Second, one thing that `ShallowWaterModel` has that the `IncompressibleModel` does not have are surface waves, so a divergence.  I think that plotting the vorticity, which is the balanced part of the flow, and the divergence (or even the free-surface), unbalanced part of the flow, might be interesting,  Or we could add the `ShallowWaterModel` as an option for the turublence example.  

Third, I don't want this to be out until we are happy with the `ShallowWaterModel`.  I am also working on the Bickley jet and hope to talk to @glw about that tomorrow to get it cleaned up.

**Comment by glwagner** [comment]
I echo @ali-ramadhan that since examples are expensive to maintain it'd be nice to have some topical diversity. Perhaps a different initial conditions in a geometry with walls, on a beta plane... ?!? We don't have any channel geometry or `BetaPlane` examples right now.

**Comment by navidcy** [comment]
Let's do shallow-water in channel geometry on a beta-plane. That'll be sweet.
(And always users can easily adjust it to get periodic domain and f=0.)

File: examples/two_dimensional_turbulence_shallow_water.jl [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    plot(
+        ω_plot, 
+        s_plot, 
+        layout = (1,2), 
+        size=(1200,500) 
+        )
+    
+end
+
+gif(anim, "two_dimensional_turbulence_shallow_water.gif", fps=8)
```
Comment (navidcy) [inline examples/two_dimensional_turbulence_shallow_water.jl:112]: change to `mp4` :)

File: examples/two_dimensional_turbulence_shallow_water.jl [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+v_op   = model.solution.vh / model.solution.h
+η_op   = model.solution.h - model.grid.Lz
+ω_op   = @at (Center, Center, Center) ∂x(v_op) - ∂y(u_op)
+s_op   = sqrt(u_op^2 + v_op^2) 
+
+u_field = ComputedField(u_op)
+v_field = ComputedField(v_op)
+η_field = ComputedField(η_op)
+ω_field = ComputedField(ω_op)
+s_field = ComputedField(s_op)
```
Comment (navidcy) [inline examples/two_dimensional_turbulence_shallow_water.jl:57]: call this `speed_field`? (similarly `speed_op`, etc?)

**Comment by glwagner** [comment]
@navidcy suggestions for a good initial condition? Forcing?

**Comment by glwagner** [comment]
Also with biharmonic diffusion?

**Comment by francispoulin** [comment]
Thanks for the feedback @glwagner and @navidcy .  I think it would be easy to do this in a channel on a beta-plane and see some zonal jets developing.  That would be great to do.  Also, if you turn off beta one should of course get a very similar result to the two-dimensional turbulence example.  I'll get to work on this today and share what I come up with.  

Do we have biharmonic diffusion in Oceananigans already?  If yes I'm also happy to play with that too.

**Comment by glwagner** [comment]
> Do we have biharmonic diffusion in Oceananigans already?

https://github.com/CliMA/Oceananigans.jl/blob/master/src/TurbulenceClosures/turbulence_closure_implementations/anisotropic_biharmonic_diffusivity.jl

Random initial conditions could be nice since that creates a illustrative contrast with the existing example (same ICs but different physics leading to dramatically different results)

File: examples/two_dimensional_turbulence_shallow_water.jl (hunk 1 above)
Comment (francispoulin) [reply examples/two_dimensional_turbulence_shallow_water.jl:112]: Good idea and happy to change it.

File: examples/two_dimensional_turbulence_shallow_water.jl (hunk 2 above)
Comment (francispoulin) [reply examples/two_dimensional_turbulence_shallow_water.jl:57]: Also a good suggestion.  Done

**Comment by francispoulin** [comment]
@Lichriszz I wanted to tell you that there is a working example of two-dimensional shallow water model in the branch `distributed-shallow-water-model`.  There is an MPI version that is being tested but the serial script should work on the main branch.

https://github.com/CliMA/Oceananigans.jl/blob/distributed-shallow-water/examples/serial_shallow_water_turbulence.jl

**Comment by francispoulin** [comment]
I don't believe there is a strong need for this example as it is.  Considering that #1430 is using two-dimensional turbulence a as a test case, and that might become an example, this PR seems a bit redundant.  That is why I will close this but happy to return to it if there is interest.  There was mention of introducing beta and looking at the development of zonal jets.  This is easy enough to do and could be fun, but not sure if this warrents an example.

**Pull request closed without merging**